File: mesh-private-messenger/apps/mobile/modules/mesh-messenger/ios/MeshMessengerModule.swift
import EXApplication
import ExpoModulesCore
import ExpoNotifications
import Foundation
import UIKit

public final class MeshMessengerModule: Module, NotificationDelegate {
  private let lock = NSLock()
  private var started = false
  private var pushEnabled = false
  private var pushPromise: Promise?

  public func definition() -> ModuleDefinition {
    Name("MeshMessenger")

    Events("onPushRegistrationChanged")

    OnCreate {
      NotificationCenterManager.shared.addDelegate(self)
      MeshMessengerDataProtection.excludeAppDataFromBackup()
      MeshMessengerDataProtection.startCoveringInactiveWindow()
    }

    AsyncFunction("primePushToken") { (promise: Promise) in
      self.pushPromise?.reject(
        "E_PUSH_REGISTRATION_REPLACED",
        "A newer push registration request replaced this one."
      )
      self.pushEnabled = true
      self.pushPromise = promise
      UIApplication.shared.registerForRemoteNotifications()
    }
    .runOnQueue(.main)

    AsyncFunction("clearPushToken") { () in
      self.pushEnabled = false
      self.pushPromise?.reject(
        "E_PUSH_REGISTRATION_CANCELLED",
        "Push registration was cancelled."
      )
      self.pushPromise = nil
      MeshMessengerClearApplePushToken()
      UIApplication.shared.unregisterForRemoteNotifications()
    }
    .runOnQueue(.main)

    AsyncFunction("invoke") { (symbol: String, request: Data) throws -> Data in
      self.lock.lock()
      defer { self.lock.unlock() }
      try self.startIfNeeded()
      switch symbol {
      case "mesh_messenger_initialize":
        return try MeshLibrary.initialize(request)
      case "mesh_messenger_validate_outer":
        return try MeshLibrary.validate_outer(request)
      case "mesh_messenger_store_envelope":
        return try MeshLibrary.persist_envelope(request)
      case "mesh_messenger_create_account":
        return try MeshLibrary.create_account_export(request)
      case "mesh_messenger_presentation_load":
        return try MeshLibrary.presentation_load_export(request)
      case "mesh_messenger_presentation_save":
        return try MeshLibrary.presentation_save_export(request)
      case "mesh_messenger_load_profile":
        return try MeshLibrary.load_profile_export(request)
      case "mesh_messenger_replenish_prekeys":
        return try MeshLibrary.replenish_prekeys_export(request)
      case "mesh_messenger_reconcile_prekeys":
        return try MeshLibrary.reconcile_prekeys_export(request)
      case "mesh_messenger_create_link_request":
        return try MeshLibrary.create_link_request_export(request)
      case "mesh_messenger_device_link_sas":
        return try MeshLibrary.device_link_sas_export(request)
      case "mesh_messenger_authorize_device_link_for_set":
        return try MeshLibrary.authorize_device_link_for_set_export(request)
      case "mesh_messenger_complete_device_link":
        return try MeshLibrary.complete_device_link_export(request)
      case "mesh_messenger_inspect_device_set":
        return try MeshLibrary.inspect_device_set_export(request)
      case "mesh_messenger_create_device_revocation":
        return try MeshLibrary.create_device_revocation_export(request)
      case "mesh_messenger_receive_initial":
        return try MeshLibrary.receive_initial_export(request)
      case "mesh_messenger_prepare_fanout_prekeys":
        return try MeshLibrary.prepare_fanout_prekeys_export(request)
      case "mesh_messenger_send_fanout":
        return try MeshLibrary.send_fanout_export(request)
      case "mesh_messenger_group_invite":
        return try MeshLibrary.group_invite_export(request)
      case "mesh_messenger_group_invitation_accept":
        return try MeshLibrary.group_invitation_accept_export(request)
      case "mesh_messenger_group_invitation_complete":
        return try MeshLibrary.group_invitation_complete_export(request)
      case "mesh_messenger_group_invitation_decline":
        return try MeshLibrary.group_invitation_decline_export(request)
      case "mesh_messenger_group_invitations":
        return try MeshLibrary.group_invitations_export(request)
      case "mesh_messenger_group_key_package":
        return try MeshLibrary.group_key_package_export(request)
      case "mesh_messenger_group_create":
        return try MeshLibrary.group_create_export(request)
      case "mesh_messenger_group_add":
        return try MeshLibrary.group_add_export(request)
      case "mesh_messenger_group_remove":
        return try MeshLibrary.group_remove_export(request)
      case "mesh_messenger_group_send":
        return try MeshLibrary.group_send_export(request)
      case "mesh_messenger_group_receive":
        return try MeshLibrary.group_receive_export(request)
      case "mesh_messenger_group_list":
        return try MeshLibrary.group_list_export(request)
      case "mesh_messenger_group_inspect":
        return try MeshLibrary.group_inspect_export(request)
      case "mesh_messenger_group_history":
        return try MeshLibrary.group_history_export(request)
      case "mesh_messenger_receive_message":
        return try MeshLibrary.receive_message_export(request)
      case "mesh_messenger_update_conversation":
        return try MeshLibrary.update_conversation_export(request)
      case "mesh_messenger_push_intent":
        return try MeshLibrary.push_intent_export(request)
      case "mesh_messenger_push_action_complete":
        return try MeshLibrary.push_action_complete_export(request)
      case "mesh_messenger_push_status":
        return try MeshLibrary.push_status_export(request)
      case "mesh_messenger_list_conversations":
        return try MeshLibrary.list_conversations_export(request)
      case "mesh_messenger_load_history":
        return try MeshLibrary.load_history_export(request)
      case "mesh_messenger_safety_number":
        return try MeshLibrary.safety_number_export(request)
      case "mesh_messenger_import_contact":
        return try MeshLibrary.import_contact_export(request)
      case "mesh_messenger_directory_entry":
        return try MeshLibrary.directory_entry_export(request)
      case "mesh_messenger_directory_lookup":
        return try MeshLibrary.directory_lookup_export(request)
      case "mesh_messenger_transparency_lookup":
        return try MeshLibrary.transparency_lookup_export(request)
      case "mesh_messenger_register_request":
        return try MeshLibrary.register_request_export(request)
      case "mesh_messenger_resolve_request":
        return try MeshLibrary.resolve_request_export(request)
      case "mesh_messenger_verify_transparency":
        return try MeshLibrary.verify_transparency_export(request)
      case "mesh_messenger_privacy_submission":
        return try MeshLibrary.privacy_submission_export(request)
      case "mesh_messenger_mailbox_fetch":
        return try MeshLibrary.mailbox_fetch_export(request)
      case "mesh_messenger_process_delivery_batch":
        return try MeshLibrary.process_delivery_batch_export(request)
      case "mesh_messenger_outbox_list":
        return try MeshLibrary.outbox_list_export(request)
      case "mesh_messenger_outbox_ack":
        return try MeshLibrary.outbox_ack_export(request)
      case "mesh_messenger_attachment_prepare":
        return try MeshLibrary.attachment_prepare_export(request)
      case "mesh_messenger_attachment_seal_chunk":
        return try MeshLibrary.attachment_seal_chunk_export(request)
      case "mesh_messenger_attachment_open_chunk":
        return try MeshLibrary.attachment_open_chunk_export(request)
      default:
        throw MeshLibraryFailure(
          status: MESH_LIBRARY_ERR_INVALID_ARGUMENT,
          payload: Data("unknown_export".utf8)
        )
      }
    }

    OnDestroy { [weak self] in
      guard let self else { return }
      NotificationCenterManager.shared.removeDelegate(self)
      MeshMessengerDataProtection.stop()
      self.pushEnabled = false
      self.pushPromise?.reject("E_MODULE_DESTROYED", "MeshMessenger was destroyed.")
      self.pushPromise = nil
      MeshMessengerClearApplePushToken()
      self.stop()
    }
  }

  public func didRegister(_ deviceToken: String) {
    guard pushEnabled else { return }
    let isRefresh = pushPromise == nil
    do {
      try cachePushToken(deviceToken)
      pushPromise?.resolve(nil)
      pushPromise = nil
      if isRefresh {
        sendEvent("onPushRegistrationChanged", [:])
      }
    } catch {
      pushPromise?.reject(error)
      pushPromise = nil
    }
  }

  public func didFailRegistration(_ error: any Error) {
    pushPromise?.reject(error)
    pushPromise = nil
  }

  private func startIfNeeded() throws {
    guard !started else { return }
    try MeshLibrary.initialize()
    let status = MeshMessengerRegisterAppleHostCallbacks()
    guard status == MESH_LIBRARY_OK else {
      throw MeshLibraryFailure(
        status: status,
        payload: Data("host_callback_registration_failed".utf8)
      )
    }
    started = true
  }

  private func stop() {
    lock.lock()
    defer { lock.unlock() }
    guard started else { return }
    MeshLibrary.shutdown()
    started = false
  }

  private func cachePushToken(_ token: String) throws {
    guard let applicationID = Bundle.main.bundleIdentifier else {
      throw PushRegistrationConfigurationError.missingApplicationID
    }
    let applicationIDBytes = Data(applicationID.utf8)
    let tokenBytes = Data(token.utf8)
    let development = EXProvisioningProfile.main().notificationServiceEnvironment() == "development"
    let status = applicationIDBytes.withUnsafeBytes { applicationIDBuffer in
      tokenBytes.withUnsafeBytes { tokenBuffer in
        MeshMessengerCacheApplePushToken(
          applicationIDBuffer.bindMemory(to: UInt8.self).baseAddress,
          UInt64(applicationIDBuffer.count),
          tokenBuffer.bindMemory(to: UInt8.self).baseAddress,
          UInt64(tokenBuffer.count),
          development
        )
      }
    }
    guard status == MESH_LIBRARY_OK else {
      throw PushRegistrationConfigurationError.invalidMaterial
    }
  }
}

private enum PushRegistrationConfigurationError: Error {
  case missingApplicationID
  case invalidMaterial
}

File: mesh-private-messenger/apps/mobile/modules/mesh-messenger/android/src/main/java/expo/modules/meshmessenger/MeshMessengerModule.kt
package expo.modules.meshmessenger

import com.google.firebase.messaging.FirebaseMessaging
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.notifications.service.delegates.FirebaseMessagingDelegate.Companion.addTokenListener
import expo.modules.notifications.service.delegates.FirebaseMessagingDelegate.Companion.removeTokenListener
import expo.modules.notifications.tokens.interfaces.FirebaseTokenListener
import java.nio.ByteBuffer
import mesh.MeshLibrary

class MeshMessengerModule : Module(), FirebaseTokenListener {
    private val lock = Any()
    private val pushLock = Any()
    private var started = false
    private var pushEnabled = false

    override fun definition() = ModuleDefinition {
        Name("MeshMessenger")

        Events("onPushRegistrationChanged")

        OnCreate {
            addTokenListener(this@MeshMessengerModule)
            MeshMessengerScreenSecurity.protect(appContext.currentActivity)
        }

        OnActivityEntersForeground {
            MeshMessengerScreenSecurity.protect(appContext.currentActivity)
        }

        AsyncFunction("primePushToken") { promise: Promise ->
            val instance = firebaseMessaging(promise) ?: return@AsyncFunction
            synchronized(pushLock) { pushEnabled = true }
            instance.isAutoInitEnabled = true
            instance.token.addOnCompleteListener { task ->
                if (!task.isSuccessful) {
                    promise.reject(
                        "E_PUSH_REGISTRATION_FAILED",
                        "Fetching the FCM token failed.",
                        task.exception,
                    )
                    return@addOnCompleteListener
                }
                val token = task.result
                if (token == null || !cachePushToken(token)) {
                    promise.reject(
                        "E_PUSH_REGISTRATION_FAILED",
                        "FCM returned invalid registration material.",
                        null,
                    )
                    return@addOnCompleteListener
                }
                promise.resolve(null)
            }
        }

        AsyncFunction("clearPushToken") { promise: Promise ->
            synchronized(pushLock) {
                pushEnabled = false
                MeshMessengerPushMaterial.clear()
            }
            val instance = firebaseMessaging(promise) ?: return@AsyncFunction
            instance.isAutoInitEnabled = false
            instance.deleteToken().addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    promise.resolve(null)
                } else {
                    promise.reject(
                        "E_PUSH_UNREGISTRATION_FAILED",
                        "Deleting the FCM token failed.",
                        task.exception,
                    )
                }
            }
        }

        AsyncFunction("invoke") { symbol: String, request: ByteArray ->
            synchronized(lock) {
                startIfNeeded()
                when (symbol) {
                    "mesh_messenger_initialize" -> MeshLibrary.initialize(request)
                    "mesh_messenger_validate_outer" -> MeshLibrary.validate_outer(request)
                    "mesh_messenger_store_envelope" -> MeshLibrary.persist_envelope(request)
                    "mesh_messenger_create_account" -> MeshLibrary.create_account_export(request)
                    "mesh_messenger_journal_load" -> MeshLibrary.journal_load_export(request)
                    "mesh_messenger_journal_save" -> MeshLibrary.journal_save_export(request)
                    "mesh_messenger_presentation_load" -> MeshLibrary.presentation_load_export(request)
                    "mesh_messenger_presentation_save" -> MeshLibrary.presentation_save_export(request)
                    "mesh_messenger_load_profile" -> MeshLibrary.load_profile_export(request)
                    "mesh_messenger_replenish_prekeys" -> MeshLibrary.replenish_prekeys_export(request)
                    "mesh_messenger_reconcile_prekeys" -> MeshLibrary.reconcile_prekeys_export(request)
                    "mesh_messenger_create_link_request" -> MeshLibrary.create_link_request_export(request)
                    "mesh_messenger_device_link_sas" -> MeshLibrary.device_link_sas_export(request)
                    "mesh_messenger_authorize_device_link_for_set" -> MeshLibrary.authorize_device_link_for_set_export(request)
                    "mesh_messenger_complete_device_link" -> MeshLibrary.complete_device_link_export(request)
                    "mesh_messenger_inspect_device_set" -> MeshLibrary.inspect_device_set_export(request)
                    "mesh_messenger_create_device_revocation" -> MeshLibrary.create_device_revocation_export(request)
                    "mesh_messenger_receive_initial" -> MeshLibrary.receive_initial_export(request)
                    "mesh_messenger_prepare_fanout_prekeys" -> MeshLibrary.prepare_fanout_prekeys_export(request)
                    "mesh_messenger_send_fanout" -> MeshLibrary.send_fanout_export(request)
                    "mesh_messenger_group_invite" -> MeshLibrary.group_invite_export(request)
                    "mesh_messenger_group_invitation_accept" -> MeshLibrary.group_invitation_accept_export(request)
                    "mesh_messenger_group_invitation_complete" -> MeshLibrary.group_invitation_complete_export(request)
                    "mesh_messenger_group_invitation_decline" -> MeshLibrary.group_invitation_decline_export(request)
                    "mesh_messenger_group_invitations" -> MeshLibrary.group_invitations_export(request)
                    "mesh_messenger_group_key_package" -> MeshLibrary.group_key_package_export(request)
                    "mesh_messenger_group_create" -> MeshLibrary.group_create_export(request)
                    "mesh_messenger_group_add" -> MeshLibrary.group_add_export(request)
                    "mesh_messenger_group_remove" -> MeshLibrary.group_remove_export(request)
                    "mesh_messenger_group_send" -> MeshLibrary.group_send_export(request)
                    "mesh_messenger_group_receive" -> MeshLibrary.group_receive_export(request)
                    "mesh_messenger_group_list" -> MeshLibrary.group_list_export(request)
                    "mesh_messenger_group_inspect" -> MeshLibrary.group_inspect_export(request)
                    "mesh_messenger_group_history" -> MeshLibrary.group_history_export(request)
                    "mesh_messenger_receive_message" -> MeshLibrary.receive_message_export(request)
                    "mesh_messenger_update_conversation" -> MeshLibrary.update_conversation_export(request)
                    "mesh_messenger_push_intent" -> MeshLibrary.push_intent_export(request)
                    "mesh_messenger_push_action_complete" -> MeshLibrary.push_action_complete_export(request)
                    "mesh_messenger_push_status" -> MeshLibrary.push_status_export(request)
                    "mesh_messenger_list_conversations" -> MeshLibrary.list_conversations_export(request)
                    "mesh_messenger_load_history" -> MeshLibrary.load_history_export(request)
                    "mesh_messenger_safety_number" -> MeshLibrary.safety_number_export(request)
                    "mesh_messenger_import_contact" -> MeshLibrary.import_contact_export(request)
                    "mesh_messenger_directory_entry" -> MeshLibrary.directory_entry_export(request)
                    "mesh_messenger_directory_lookup" -> MeshLibrary.directory_lookup_export(request)
                    "mesh_messenger_transparency_lookup" -> MeshLibrary.transparency_lookup_export(request)
                    "mesh_messenger_register_request" -> MeshLibrary.register_request_export(request)
                    "mesh_messenger_resolve_request" -> MeshLibrary.resolve_request_export(request)
                    "mesh_messenger_verify_transparency" -> MeshLibrary.verify_transparency_export(request)
                    "mesh_messenger_privacy_submission" -> MeshLibrary.privacy_submission_export(request)
                    "mesh_messenger_mailbox_fetch" -> MeshLibrary.mailbox_fetch_export(request)
                    "mesh_messenger_process_delivery_batch" -> MeshLibrary.process_delivery_batch_export(request)
                    "mesh_messenger_outbox_list" -> MeshLibrary.outbox_list_export(request)
                    "mesh_messenger_outbox_ack" -> MeshLibrary.outbox_ack_export(request)
                    "mesh_messenger_outbox_fail" -> MeshLibrary.outbox_fail_export(request)
                    "mesh_messenger_outbox_page" -> MeshLibrary.outbox_page_export(request)
                    "mesh_messenger_attachment_prepare" -> MeshLibrary.attachment_prepare_export(request)
                    "mesh_messenger_attachment_seal_chunk" -> MeshLibrary.attachment_seal_chunk_export(request)
                    "mesh_messenger_attachment_open_chunk" -> MeshLibrary.attachment_open_chunk_export(request)
                    else -> throw IllegalArgumentException("unknown_export")
                }
            }
        }

        OnDestroy {
            removeTokenListener(this@MeshMessengerModule)
            synchronized(pushLock) {
                pushEnabled = false
                MeshMessengerPushMaterial.clear()
            }
            synchronized(lock) {
                if (started) {
                    MeshMessengerHost.unregisterHostCallbacks()
                    MeshLibrary.shutdownNative()
                    started = false
                }
            }
        }
    }

    override fun onNewToken(token: String) {
        if (cachePushToken(token)) {
            runCatching { sendEvent("onPushRegistrationChanged", emptyMap<String, Any?>()) }
        }
    }

    private fun startIfNeeded() {
        if (started) return
        val context = appContext.reactContext
            ?: throw IllegalStateException("React context is unavailable")
        MeshMessengerSecureStore.install(context)
        MeshMessengerPushMaterial.install(context)
        MeshLibrary.ensureInitialized()
        val status = MeshMessengerHost.registerHostCallbacks()
        check(status == 0) { "Host callback registration failed (status=$status)" }
        started = true
    }

    private fun cachePushToken(token: String): Boolean = synchronized(pushLock) {
        if (!pushEnabled) return@synchronized false
        val applicationID = appContext.reactContext?.applicationContext?.packageName
            ?: return@synchronized false
        MeshMessengerPushMaterial.cache(applicationID, token)
    }

    private fun firebaseMessaging(promise: Promise): FirebaseMessaging? = try {
        FirebaseMessaging.getInstance()
    } catch (error: IllegalStateException) {
        promise.reject(
            "E_PUSH_REGISTRATION_FAILED",
            "Firebase Messaging is not configured.",
            error,
        )
        null
    }
}

internal object MeshMessengerHost {
    @JvmStatic external fun registerHostCallbacks(): Int
    @JvmStatic external fun unregisterHostCallbacks()
}

internal object MeshMessengerPushMaterial {
    private const val PROJECT_ID_METADATA = "app.morse.mesh.EXPO_PROJECT_ID"
    private const val BROKER_KEY_METADATA = "app.morse.mesh.PUSH_BROKER_PUBLIC_KEY_HEX"
    private const val SECURITY_CONFIG_METADATA = "app.morse.mesh.SECURITY_CONFIG"
    private const val MAX_APPLICATION_ID_BYTES = 255
    private const val MAX_TOKEN_BYTES = 4096
    private const val MAX_FRAME_BYTES = 4362
    private var frame: ByteArray? = null
    private var buildConfigFrame: ByteArray? = null
    private var securityConfigFrame: ByteArray? = null
    private var buildConfigLoaded = false

    @Suppress("DEPRECATION")
    @Synchronized
    fun install(context: android.content.Context) {
        if (buildConfigLoaded) return
        val metadata = context.packageManager
            .getApplicationInfo(context.packageName, android.content.pm.PackageManager.GET_META_DATA)
            .metaData
        val projectID = metadata?.get(PROJECT_ID_METADATA)?.toString()
        val brokerPublicKeyHex = metadata?.get(BROKER_KEY_METADATA)?.toString()
        buildConfigFrame = if (projectID == null && brokerPublicKeyHex == null) {
            null
        } else {
            "1\n${projectID.orEmpty()}\n${brokerPublicKeyHex.orEmpty()}".toByteArray(Charsets.UTF_8)
        }
        securityConfigFrame = metadata?.get(SECURITY_CONFIG_METADATA)?.toString()
            ?.toByteArray(Charsets.UTF_8)
        buildConfigLoaded = true
    }

    @Synchronized
    fun cache(applicationID: String, token: String): Boolean {
        val applicationIDBytes = applicationID.toByteArray(Charsets.UTF_8)
        val tokenBytes = token.toByteArray(Charsets.UTF_8)
        if (applicationIDBytes.size !in 1..MAX_APPLICATION_ID_BYTES ||
            tokenBytes.size !in 1..MAX_TOKEN_BYTES
        ) {
            tokenBytes.fill(0)
            return false
        }
        val size = 11 + applicationIDBytes.size + tokenBytes.size
        if (size > MAX_FRAME_BYTES) {
            tokenBytes.fill(0)
            return false
        }
        val replacement = ByteBuffer.allocate(size)
            .put(1.toByte())
            .put(2.toByte())
            .put(0.toByte())
            .putInt(applicationIDBytes.size)
            .put(applicationIDBytes)
            .putInt(tokenBytes.size)
            .put(tokenBytes)
            .array()
        tokenBytes.fill(0)
        frame?.fill(0)
        frame = replacement
        return true
    }

    @JvmStatic
    @Synchronized
    fun consume(): ByteArray? = frame.also { frame = null }

    @JvmStatic
    @Synchronized
    fun buildConfig(): ByteArray? = buildConfigFrame?.clone()

    @JvmStatic
    @Synchronized
    fun securityConfig(): ByteArray? = securityConfigFrame?.clone()

    @Synchronized
    fun clear() {
        frame?.fill(0)
        frame = null
    }
}
